File: solutions/cpp/0297.cpp
class Codec {
 public:
  string serialize(TreeNode* root) {
    string ans;

    function<void(TreeNode*)> helper = [&](TreeNode* root) {
      if (!root) {
        ans += "# ";
        return;
      }

      ans += to_string(root->val) + " ";
      helper(root->left);
      helper(root->right);
    };

    helper(root);

    return ans;
  }

  TreeNode* deserialize(string data) {
    istringstream iss(data);
    queue<string> q;

    for (string s; iss >> s;)
      q.push(s);

    function<TreeNode*()> helper = [&]() -> TreeNode* {
      const string s = q.front(); q.pop();
      if (s == "#") return nullptr;

      TreeNode* root = new TreeNode(stoi(s));
      root->left = helper();
      root->right = helper();

      return root;
    };

    return helper();
  }
};

File: solutions/java/0297.java
public class Codec {
  public String serialize(TreeNode root) {
    if (root == null)
      return "";

    StringBuilder ans = new StringBuilder();
    Queue<TreeNode> queue = new LinkedList<>();
    queue.add(root);

    while (!queue.isEmpty()) {
      TreeNode node = queue.poll();
      if (node == null) {
        ans.append("n ");
        continue;
      }
      ans.append(node.val + " ");
      queue.add(node.left);
      queue.add(node.right);
    }

    return ans.toString();
  }

  public TreeNode deserialize(String data) {
    if (data.equals(""))
      return null;

    String[] vals = data.split(" ");
    TreeNode root = new TreeNode(Integer.parseInt(vals[0]));
    Queue<TreeNode> queue = new LinkedList<>();
    queue.add(root);

    for (int i = 1; i < vals.length; ++i) {
      TreeNode parent = queue.poll();
      if (!vals[i].equals("n")) {
        TreeNode left = new TreeNode(Integer.parseInt(vals[i]));
        parent.left = left;
        queue.add(left);
      }
      if (!vals[++i].equals("n")) {
        TreeNode right = new TreeNode(Integer.parseInt(vals[i]));
        parent.right = right;
        queue.add(right);
      }
    }

    return root;
  }
}